### PR TITLE
COL-43 - Added Apache configuration and deploy script

### PR DESCRIPTION
**Pull request by nicolaasmatthijs**
https://jira.ets.berkeley.edu/jira/browse/COL-43

- Apache configuration generator script
- Deploy script
- Removed cors package in favor of Apache
- Fixes to using LTI tools over https behind Apache

### CODE REVIEW

**Comment by nicolaasmatthijs** [comment]
@simong : Would you be able to have a thorough look at this?

**Comment by nicolaasmatthijs** [comment]
Note: this is not entirely ready for merging yet as I'd still like to add some additional documentation

File: apache/apache.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+var _ = require('lodash');
+var config = require('config');
+var fs = require('fs');
+
+// Read the Apache config template
+fs.readFile(__dirname + '/collabosphere.template', 'utf8', function (err,template) {
+  if (err) {
+    console.log('An error occurred when reading the Apache config template');
```
Comment (simong) [inline apache/apache.js]: You probably want to return or `process.exit(1)` here?

File: apache/apache.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+var _ = require('lodash');
+var config = require('config');
+var fs = require('fs');
+
+// Read the Apache config template
+fs.readFile(__dirname + '/collabosphere.template', 'utf8', function (err,template) {
```
Comment (simong) [inline apache/apache.js]: Should the lack of space between `,` and `template` not trip up the code style checker?

File: apache/apache.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    'logDirectory': config.get('apache.logDirectory'),
+    'port': config.get('app.port')
+  };
+
+  var templateOutput = _.template(template)(templateData);
+
+  // Store the generated Apache config file
+  fs.writeFile(__dirname + '/collabosphere.conf', templateOutput, function (err) {
+    if (err) {
+      console.log('An error occurred when writing the generated Apache config file');
```
Comment (simong) [inline apache/apache.js]: return or `process.exit(1)`

File: apache/collabosphere.template [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+<VirtualHost *:80>
+
+    DocumentRoot "<%= documentRoot %>"
+
+    Header set Access-Control-Allow-Origin "*"
+
+    ErrorLog <%= logDirectory %>/collabosphere_error.log
+    CustomLog <%= logDirectory %>/collabosphere_access_log common
+    CustomLog <%= logDirectory %>/collabosphere-combined_log combined
```
Comment (simong) [inline apache/collabosphere.template]: Is the `combined` log necessary if we already have a `common` logger?

File: apache/collabosphere.template [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+    # Proxy API & LTI calls to the application server
+    ProxyPass /api http://localhost:<%= port %>/api retry=0
+    ProxyPass /lti http://localhost:<%= port %>/lti retry=0
+    ProxyPreserveHost On
+
+    RewriteEngine On
+    RewriteRule ^/assetlibrary         <%= documentRoot %>/index.html
+
+</VirtualHost>
```
Comment (simong) [inline apache/collabosphere.template]: Missing line-break

File: package.json [hunk 6]
```diff
@@ -19,11 +19,18 @@
...
+    "col-lti": "latest",
+    "col-rest": "latest",
+    "col-tests": "latest",
+    "col-users": "latest",
     "config": "latest",
     "cookie-parser": "latest",
-    "cors": "latest",
     "express": "latest",
-    "ims-lti": "latest",
+    "ims-lti": "git://github.com/nicolaasmatthijs/ims-lti.git",
```
Comment (simong) [inline package.json]: A PR from you `master` branch? Naughty, naughty.. ^^

**Comment by simong** [comment]
This PR LGTM, couple of minor styling issues. You might want to document somewhere that Apache needs `RequestHeader set X-Forwarded-Proto "https"` when serving through SSL.

File: apache/apache.js (hunk 2 above)
Comment (nicolaasmatthijs) [inline apache/apache.js]: I've added the apache config generator script to the list of files jscs should check

File: apache/collabosphere.template (hunk 4 above)
Comment (nicolaasmatthijs) [inline apache/collabosphere.template]: This is desired practice, as used in Data Cultures

**Comment by nicolaasmatthijs** [comment]
Followed up on review. Back to @simong for further review and merging

File: package.json [hunk 7]
```diff
@@ -19,6 +19,15 @@
     "body-parser": "latest",
     "bunyan": "latest",
     "bunyan-prettystream": "latest",
+    "clone": "latest",
```
Comment (simong) [inline package.json]: Is it really necessary to bring in `clone`?  Lodash has a `_.cloneDeep` function that does a lot of what `clone` does.

**Pull request merged**